Codes for ODLA_POPART pipeline supports

jackzipu opened the pull request:
Very beginning of codes with some hard coded codes to run mlperf model on 1 IPU in BS6 and got the expected result.

jackzipu committed:
First runnable with single thread & test context

jackzipu committed:
mnist runnable demot to test the pipeline

jackzipu committed:
multi thread put the data to the session run

jackzipu committed:
simple bash to compile and run test

jackzipu committed:
An example of how to use the callback in pipeline

jackzipu committed:
multi threads using local Ctx

jackzipu committed:
Can run with pipeline setting in onnx file

jackzipu committed:
Refactored and add no pipeline multi thread

jackzipu committed:
Move codes to the odla_pipeline.h .cc

jackzipu committed:
Make single empty/zero data, and delete context for empty data after get result

jackzipu committed:
Add mutex to serialization the compute requests

jackzipu committed:
Merge branch 'master' of github.com:jackzipu/heterogeneity-aware-lowering-and-optimization

jackzipu committed:
merge from upstream master

jackzipu committed:
Merge the changes for attention mask & prevous changes

jackzipu committed:
test codes for time

jackzipu committed:
Chage the CMakeList to make the pipeline.cc and new custom op compiled

jackzipu committed:
Successfully run on 24L with attention mask custom OP

jackzipu committed:
custom op attention_mask test code

jackzipu committed:
And name scope to the each node in model

jackzipu committed:
Try throghput test with MLPerf model

jackzipu committed:
only set AMP on feed forward matmul

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/alibaba/heterogeneity-aware-lowering-and-optimization?pullRequest=417) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/alibaba/heterogeneity-aware-lowering-and-optimization?pullRequest=417) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: jackzipu<br/>:x: gcuser<br/><hr/>**gcuser** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/alibaba/heterogeneity-aware-lowering-and-optimization?pullRequest=417) it.</sub>

jackzipu committed:
Run the online pipeling with config hard coded to the config read class

jackzipu committed:
Compile with SDK 2.2 with pipeline online setting

jackzipu committed:
Add config file for pipeline stage setting

jackzipu committed:
Run pipeline with similar performance of popart

jackzipu committed:
change some names & make AMP all 0.445

jackzipu committed:
Add amp parameter in config file

jackzipu committed:
Detach device and clear session when DestroyComputation

jackzipu committed:
Make the batch_per_step take effect on execution mode SEQUENCE to pass enough size of data

jackzipu committed:
Add the new lock free queue and logging

jackzipu committed:
Fix bug on empty data visit counter

jackzipu committed:
delete the empty context

jackzipu committed:
add some pipeline sync

jackzipu committed:
Make thread sleep for 5 ms when no task in the queue

jackzipu committed:
change the size() of LockFreeQueue to tail-wait

jackzipu committed:
[CI] make the call by main can work with npz files

jackzipu committed:
Move the computation init to create context

jackzipu committed:
Add common functions to common.h and common.cc

jackzipu committed:
move the compuation init out

Commit:
[CI] factor out checkout job

jackzipu committed:
Move common functions to the test foler

jackzipu committed:
Merge form alibaba/halo master to align the codes

jackzipu committed:
Test the config of ODLA popART and make no configuration act as before

jackzipu committed:
Add tests for call the model.cc

jackzipu committed:
Add FP32 to save as result

Commit:
Some changes on LockFreeQueue and tests

jackzipu committed:
Fix the rsqrt wrong problem, and remove std cout&cerr to avoid crash

jackzipu committed:
fix the accuracy problem of large bps

jackzipu committed:
Add thread check for context & computation holding to avoid conflicts

jackzipu committed:
Add the batch tools to help on the test to generate model, build and run

youbeny commented on ODLA/platforms/odla_popart/odla_compute.cc:
```diff
@@ -94,210 +58,236 @@ odla_status odla_SetComputationItem(odla_computation comp, odla_item_type type,
+    return ODLA_FAILURE;
+  comp->executor()->compute(comp, context, mode, device);
+  //std::cout << "<--- odla_ExecuteComputation()" << std::endl;
   return ODLA_SUCCESS;
```
pipeline mode needs to be compatible with non-pipeline mode，consider add a flag， such as  
```cpp
if (comp->opts.enable_pipeline) {
        // do pipeline;
} else {
        // normal;
}
```

youbeny commented on ODLA/platforms/odla_popart/odla_popart.h:
```diff
@@ -58,31 +71,77 @@ struct _odla_value {
+  std::mutex m_init_mutex;
+  Execution* m_executor;
+  std::thread::id thread_id_of_holder;
+
```
mixed code style  m_xx / xx_ / _xx

jackzipu committed:
Decreasing the empty data put

Commit:
temporary commit to migrate crashed system

jackzipu committed:
set pipeline information on fly
change the mixed style of class member
add debug setting and default to false to make the opts set by api
remove the old pipeline set api

jackzipu committed:
Fixed the mixed code style and removed redundant codes

jackzipu committed:
Remove the function test codes of the odla_popart

jackzipu committed:
remove some redundant codes and files

weimingzha0 commented:
where is ODLA/platforms/odla_popart/json.hpp from? Is it some third-party code?

weimingzha0 commented on ODLA/platforms/odla_popart/odla_popart.h:
```diff
@@ -20,29 +20,42 @@
-#include <string>
-#include <vector>
+
+#define g_comp _odla_computation::instance()
```
can we use "static _odla_computation* &g_comp =  _odla_computation::instance_" ?

jackzipu replied:
We need the instance() function to do extra check of the thread id which is holding the computation object, like follows
`static _odla_computation* instance(bool hold_it = true){
      if(hold_it)
          instance_->hold(); 
      return instance_;
  }`

weimingzha0 commented:
can we support multiple computation on different IPU?

jackzipu commented on ODLA/platforms/odla_popart/CMakeLists.txt:
```diff
@@ -51,3 +49,8 @@ if (NOT ODLA_BUILD_POPART_USE_CXX11ABI)
 
 target_link_libraries(odla_popart PUBLIC ODLA custom_ops popart-only)
+
+target_include_directories(odla_popart PRIVATE
```
We need this as the functions of pipeline stage setting need to operate the onnx model, this function is in odla_popart.cc, which compiled by this CMakeList.

jackzipu commented on ODLA/platforms/odla_popart/CMakeLists.txt:
```diff
@@ -20,17 +20,15 @@ option(ODLA_BUILD_POPART_CUSTOM_OPS "Link with Popart custom ops" ON)
+                               )
 if (NOT POPLAR_ROOT)
-  set(POPLAR_ROOT "/opt/poplar_sdk/poplar" CACHE PATH "Path of poplar root")
+  set(POPLAR_ROOT "/opt/poplar_sdk/poplar" CACHE STRING "Path of poplar root")
```
Changed to PATH for this.

jackzipu reviewed (commented):
> where is ODLA/platforms/odla_popart/json.hpp from? Is it some third-party code?

Yes, it's a json parser of third party to parse the pipeline configuration.  Only this one hpp codes needed to be included.

jackzipu commented:
> can we support multiple computation on different IPU?
In current design & architecture, I'm afraid we can't.  You can see that all the OPs called without passed in odla_computation object or pointer.  We used a g_comp, which meant only 1 graph can be used at same time in one process, otherwise we could not tell which computation the Ops belonged.

For the SDK itself whether support multiple graphs with multi-threads, we can have a fast check if needed.  Thanks.

jackzipu committed:
Changed the CACHE STRING to CACHE PATH

weimingzha0 commented:
> > where is ODLA/platforms/odla_popart/json.hpp from? Is it some third-party code?
> 
> Yes, it's a json parser of third party to parse the pipeline configuration. Only this one hpp codes needed to be included.

can we put it in a "third_party" directory? (e.g.  ODLA/platforms/odla_popart/custom_ops/third_party )

jackzipu commented:
The code has been merged by pull request 522, close this PR.
